softlayer/resource_softlayer_bare_metal.go

```diff
@@ -0,0 +1,442 @@
+	if err != nil {
+		return fmt.Errorf("Error deleting bare metal server while waiting for zero active transactions: %s", err)
+	}
+
+	billingItem, err := service.Id(id).GetBillingItem()
+	if err != nil {
+		return fmt.Errorf("Error getting billing item for bare metal server: %s", err)
+	}
+
+	billingItemService := services.GetBillingItemService(sess)
+	_, err = billingItemService.Id(*billingItem.Id).CancelItem(
+		sl.Bool(true), sl.Bool(true), sl.String("No longer required"), sl.String("Please cancel this server"),
+	)
+	if err != nil {
+		return fmt.Errorf("Error canceling the bare metal server (%d): %s", id, err)
+	}
+
+	return nil
+}
+
+func resourceSoftLayerBareMetalExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	service := services.GetHardwareService(meta.(*session.Session))
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return false, fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	result, err := service.Id(id).GetObject()
+	if err != nil {
+		if apiErr, ok := err.(sl.Error); !ok || apiErr.StatusCode != 404 {
+			return false, fmt.Errorf("Error trying to retrieve the Bare Metal server: %s", err)
+		}
+	}
+
+	return err == nil && result.Id != nil && *result.Id == id, nil
+}
+
+// Bare metal creation does not return a bare metal object with an Id.
+// Have to wait on provision date to become available on server that matches
+// hostname and domain.
+// http://sldn.softlayer.com/blog/bpotter/ordering-bare-metal-servers-using-softlayer-api
+func WaitForBareMetalProvision(d *datatypes.Hardware, meta interface{}) (interface{}, error) {
+	hostname := *d.Hostname
+	domain := *d.Domain
+	log.Printf("Waiting for server (%s.%s) to have to be provisioned", hostname, domain)
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"", "pending"},
+		Target:  []string{"provisioned"},
+		Refresh: func() (interface{}, string, error) {
+			service := services.GetAccountService(meta.(*session.Session))
+			bms, err := service.Filter(
+				filter.Build(
+					filter.Path("hardware.hostname").Eq(hostname),
+					filter.Path("hardware.domain").Eq(domain),
+				),
+			).Mask("id,provisionDate").GetHardware()
+			if err != nil {
+				return nil, "", fmt.Errorf("Problem fetching bare metal servers matching %s.%s: %s", hostname, domain, err)
+			}
+
+			if len(bms) == 0 || bms[0].ProvisionDate == nil {
+				return datatypes.Hardware{}, "pending", nil
+			} else {
+				return bms[0], "provisioned", nil
+			}
+		},
+		Timeout:    4 * time.Hour,
+		Delay:      30 * time.Second,
+		MinTimeout: 2 * time.Minute,
+	}
+
+	return stateConf.WaitForState()
+}
+
+func WaitForNoBareMetalActiveTransactions(id int, meta interface{}) (interface{}, error) {
+	log.Printf("Waiting for server (%d) to have zero active transactions", id)
+	service := services.GetHardwareServerService(meta.(*session.Session))
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"", "active"},
```

Comment:
I only see `""` returned when there is an error retrieving the object.  But since the refresh func returns the error, I don't think this state is ever checked:

```go
// ... in WaitForState()
// If the Refresh function returns a error, exit immediately with that error.
```

net: I don't think you need `""` here.  Or, if you intend to retry in the event of an error, then don't return the error, and return a state listed here (`""` or `"pending"`